### ISSUE TITLE
Doubts about training

### ISSUE THREAD

**Issue by TheMrguiller** [dhansmair/flamingo-mini#17]
Hi,

I am new to multi modal models and I have several questions regarding the training process. First of all, how do you feed the model with the training data? Do you insert as tokens the features of the images or do you feed the model with the new tokens created in the presampler?Is it the input the image and the label the text, or do you insert some extra text in the input?Secondly, what kind of hardware have you use to train the model and what was the time needed to train it? Lastly, what are the main differences of this project in comparison with the one published by Deepmind´s Flamingo?

Thank you in advanced

**Comment by dhansmair** [comment]
Hi @TheMrguiller, 

The forward pass works like this:
Given an image and a caption, the image is fed to the vision encoder (ViT) and perceiver resampler, obtaining a downscaled set of visual features.
The caption is transformed into the input sequence `"<BOS><image>{caption}<EOC><EOS>"`. Note here that BOS, EOC and EOS are tokens, while the `<image>` tag is inserted in plain text in accordance with deepminds flamingo.
Next, the input sequence is fed through the LM, while the cross-attention layers attend to the visual features.
The model is trained using the causal language modeling objective.

I don't remember the exact training settings for the models that I uploaded. It was a single GPU with ~40GB of memory and trained for around 60h if I remember correctly. The tiny model only occupies around 4GB on the GPU, the mini model is a bit larger.

Regarding your last question: Besides the obvious differences of tokenizer, pretrained LM and vision encoder, training data and general scale, I tried to follow the details outlined in the paper as close as possible. But I can't give any guarantees since I don't know their code.

Best, David

**Comment by TheMrguiller** [comment]
Thanks you so much for your response. Regarding the input sequence, you dont feed any labels to the model? I am used to train seqtoseq and classification task, so its really new to me. Did you use all the COCO dataset for the training?
Thank you in advanced

**Comment by dhansmair** [comment]
For the models I uploaded, I didn't use COCO but Conceptual Captions. But anyway, the training data is image-text pairs.
Maybe I should elaborate on that, because you are mentioning 'labels': COCO has labels for classification and segmentation, but it also has image captions, which are textual descriptions of the scene. So it's not a classification task, but a text generation task.
The models are trained with self-supervised learning, which means that both the image and the text are fed to the model.

Does that answer your question?

**Comment by TheMrguiller** [comment]
I am so sorry for the missunderstood. My question was related to the input, let me put an example. Image you have an image of a dog running and the caption says a corgi running in a red suit. The input of the model should be the following: `"<BOS><image><EOS>"`  and the label: " a corgi running in a red suit" ? Or is it just the input al together like you mentioned: `"<BOS><image>{caption}<EOC><EOS>"` . My question was related to the structured of the inputs and label.

**Comment by dhansmair** [comment]
Yeah it's exactly like that, the input in your case would be `"<BOS><image>a corgi running in a red suit<EOC><EOS>"`

**Comment by TheMrguiller** [comment]
I am sorry for bothering you, but as i am using as a base for a university project I am trying to get the grasp of it. I have several questions for you, i am a newbie in nlp and in multi modal:

1. Why do you randomly select a caption while training the model instead of using the correct caption of the data.
2. Is it possible to create a model of visual question and answering for your model based on the training you did for the tiny version? As far as i understand, it seems feasible, It just need some changes on the preprocess function and and that's it.
3. I don't really get how you use the CocoCaptions in your model. It seems that you process the images and  text separately, but i don't get the structure. I will check it with a small dataset, as i a, going to use my own dataset.
4.Could you upload the training parameters used for the mini model? it will be great ^^.

Thank you so much for your responses, i was so lost due to lack of knowledge.

**Comment by dhansmair** [comment]
hi, no worries, it's nice to hear you want to use it! regarding your questions:
1. coco has five captions per image, however during training I only need one sentence. As they are all describing the same image and the content of the captions should be quite similar, I just do this as a kind of data augmentation, as you would also do in computer vision
2. yes, that should be possible. But no guarantees about the performance ;) the model is small for a language model
3. I checked the code again, have to admit it is confusing. I did the image and text preprocessing in the dataset and the tokenization in the data collator. 
If I remember correctly, I did the tokenization in the data collator because this way, the input sequences in the batch are automatically padded to the same length. If you do the tokenization in the dataset, you will still need to do the padding manually in the collator. Another alternative might be to use https://huggingface.co/docs/datasets/, they have a different api but I did not have the time to dig into it.
4. not sure atm, I might check tomorrow

**Comment by TheMrguiller** [comment]
I am really grateful with your answers and your speed, I think you have been one of the most helpful people so far to solve problems in Github. Our objective in these case is to learn how these models works and how to train and finetune them, as its only an introduction to NLP, and you have enough comments to understand several complex subjects. Maybe i would ask you for some advised in the near feature as we are trying to develop and visual question and answering and chain of thought model.

Again i really appreciate your help and efforts. Great dude!!

**Comment by TheMrguiller** [comment]
Sorry for bothering you, but we are having problems adapting to question answering as it gives a size mismatch problem.
```
File "/home/d4k/anaconda3/envs/flamingo/lib/python3.10/site-packages/flamingo_mini_task/modeling_flamingo.py", line 297, in forward
    loss = F.cross_entropy(shift_logits.view(-1, shift_logits.size(-1)),
  File "/home/d4k/anaconda3/envs/flamingo/lib/python3.10/site-packages/torch/nn/functional.py", line 3026, in cross_entropy
    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
ValueError: Expected input batch_size (672) to match target batch_size (96).
```
Could you give us some advice? I know that the loss function you use is for predicting the next word in the sentence, and this problem is quite diferent. It is more related to input and output match.

**Comment by dhansmair** [comment]
I don't know what you are doing, could you list the shapes of the input data you feed to the model?

my guess is you construct sentences of kind `<image>question: {question} answer: {anwer}<EOC>` or sth similar? Because flamingo uses a decoder LM and not an encoder-decoder

**Comment by TheMrguiller** [comment]
We are trying to train in on Chain of Thought and Question answering given multiple choice. The input is shaped like these:
```
if data["solution"]!="":
            return f"{random.choice(['', ' '])}[COT][CONTEXT]<image>{data['question']}{data['choices']}</s>"
        else:
            return f"{random.choice(['', ' '])}[QA][CONTEXT]<image>{data['question']}{data['choices']}</s>"
```
Depending on the task the output should look like the following:
- "metal"
- "Look at the picture of the door. This door is made of two different materials. The body of the door is made of wood. The door knob is made of metal. Not all doors are made of wood. The body of a door can be made of metal or even glass."

The thing is that the output size doesnt match the input size. I tried adding padding but i dont know if it is the right approach
